### PR TITLE
Feat background task nf

### PR DESCRIPTION
**Pull request by nazarfil**
Creates dataset snapshot model and the functional flow to generate snapshot object in the task
Exposes snapshot object through GraphQL api

### CODE REVIEW

**Comment by nazarfil** [comment]
wip: adding tests

File: hexa/datasets/queue.py [hunk 1]
```diff
@@ -1,18 +1,81 @@
...
+            )
+            logger.info(f"Dataset snapshot saved for file {dataset_version_file_id}")
+        else:
+            logger.info(f"Dataset snapshot is empty for file {dataset_version_file_id}")
+        dataset_file_metadata.status = DatasetFileMetadata.STATUS_FINISHED
+        dataset_file_metadata.save()
+        logger.info("Dataset snapshot created for file {dataset_version_file_id}")
+    except Exception as e:
+        dataset_version_file_id = job.args["file_id"]
+        dataset_version_file = DatasetVersionFile.objects.get(
```
Comment (qgerome) [inline hexa/datasets/queue.py:68]: Don't you have the `dataset_version_file` already ?

File: hexa/datasets/queue.py [hunk 2]
```diff
@@ -1,18 +1,81 @@
...
+            dataset_file_metadata.content = file_snapshot_content.to_json(
+                orient="records"
+            )
+            logger.info(f"Dataset snapshot saved for file {dataset_version_file_id}")
+        else:
+            logger.info(f"Dataset snapshot is empty for file {dataset_version_file_id}")
+        dataset_file_metadata.status = DatasetFileMetadata.STATUS_FINISHED
+        dataset_file_metadata.save()
+        logger.info("Dataset snapshot created for file {dataset_version_file_id}")
+    except Exception as e:
```
Comment (qgerome) [inline hexa/datasets/queue.py:66]: This whole block is to be reviewed. You do not know why it raised an error but you still try to get the DatasetVersionFile & the DatasetFileMetadata.

Comment (qgerome) [reply hexa/datasets/queue.py:66]: You have to distinguish the `DatesetVersionFile.DoesNotExist` from the rest as a start

File: hexa/datasets/queue.py [hunk 3]
```diff
@@ -1,18 +1,81 @@
...
+            dataset_version_file=dataset_version_file,
+            status=DatasetFileMetadata.STATUS_PROCESSING,
+        )
+
+        download_url = generate_download_url(dataset_version_file)
+        file_snapshot_df = read_file_content(
+            download_url, dataset_version_file.content_type
+        )
+        if not file_snapshot_df.empty:
+            file_snapshot_content = file_snapshot_df.head(
```
Comment (qgerome) [inline hexa/datasets/queue.py:54]: Not `.head` but `.sample`
(cfr https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.sample.html)

File: hexa/datasets/queue.py [hunk 4]
```diff
@@ -1,18 +1,81 @@
...
+from hexa.datasets.models import (
+    DatasetFileMetadata,
+    DatasetFileMetadataJob,
+    DatasetVersionFile,
+)
 
 logger = getLogger(__name__)
 
 
+def read_file_content(download_url: str, content_type: str) -> pd.DataFrame:
```
Comment (qgerome) [inline hexa/datasets/queue.py:17]: I do not think we need this function or at least to directly use the `DatasetVersionFile`

File: hexa/datasets/queue.py [hunk 5]
```diff
@@ -1,18 +1,81 @@
...
+        dataset_file_metadata = DatasetFileMetadata.objects.create(
+            dataset_version_file=dataset_version_file,
+            status=DatasetFileMetadata.STATUS_PROCESSING,
+        )
+
+        download_url = generate_download_url(dataset_version_file)
+        file_snapshot_df = read_file_content(
+            download_url, dataset_version_file.content_type
+        )
+        if not file_snapshot_df.empty:
```
Comment (qgerome) [inline hexa/datasets/queue.py:53]: What if the file is empty? We could maybe set the sample to [] no ?

File: hexa/datasets/queue.py [hunk 6]
```diff
@@ -1,18 +1,81 @@
...
+        )
+        logger.info(
+            f"Creating dataset snapshot for version file {dataset_version_file_id}"
+        )
+        dataset_file_metadata = DatasetFileMetadata.objects.create(
+            dataset_version_file=dataset_version_file,
+            status=DatasetFileMetadata.STATUS_PROCESSING,
+        )
+
+        download_url = generate_download_url(dataset_version_file)
```
Comment (qgerome) [inline hexa/datasets/queue.py:49]: This could be in the `read_file_content` function.

File: hexa/datasets/queue.py (hunk 4 above)
Comment (qgerome) [reply hexa/datasets/queue.py:17]: And maybe rename it to `get_dataframe` or something more self-explanatory to what your returns no ?

File: hexa/datasets/schema/queries.py [hunk 7]
```diff
@@ -37,6 +42,21 @@ def resolve_dataset_version(_, info, **kwargs):
         return None
 
 
+@datasets_queries.field("datasetFileSnapshot")
+def resolve_dataset_file_snapshot(_, info, **kwargs):
+    try:
+        if kwargs.get("file_id"):
+            return DatasetFileMetadata.objects.get(
```
Comment (qgerome) [inline hexa/datasets/schema/queries.py:49]: I'm pretty sure that you have multiple `DatasetFileMetadata` per `DatasetVersionFile`

File: hexa/datasets/queue.py [hunk 8]
```diff
@@ -1,18 +1,81 @@
...
+)
 
 logger = getLogger(__name__)
 
 
+def read_file_content(download_url: str, content_type: str) -> pd.DataFrame:
+    try:
+        if content_type == "text/csv":
+            return pd.read_csv(download_url)
+        elif content_type == "application/octet-stream":
```
[the remark this answers is not in the document]
Comment (nazarfil) [reply hexa/datasets/queue.py:21]: Yeah, ill check on filename then

File: hexa/datasets/queue.py (hunk 3 above)
Comment (nazarfil) [reply hexa/datasets/queue.py:54]: Head works for files that have less lines then requested, and always gives same ordered result, which works well with assertion. I would still keep head

File: hexa/datasets/queue.py (hunk 2 above)
Comment (nazarfil) [reply hexa/datasets/queue.py:66]: Agree yeah I will rewrite together with covering edge cases in tests

File: hexa/datasets/schema/queries.py (hunk 7 above)
Comment (nazarfil) [reply hexa/datasets/schema/queries.py:49]: I was in doubt for this one as it would make sense as well to update the content of the sample instead of creating  a new one.  Would one make use of a history of file samples?

File: hexa/datasets/queue.py (hunk 1 above)
Comment (nazarfil) [reply hexa/datasets/queue.py:68]: As it was in try block I have requested it in here, but indeed could be put requested once for both.

File: hexa/datasets/graphql/schema.graphql [hunk 9]
```diff
@@ -424,11 +424,20 @@ type PinDatasetResult {
   errors: [PinDatasetError!]!
 }
 
+type DatasetFileMetadata {
+    content: JSON
+    status: String!
+    datasetVersionFile: DatasetVersionFile
```
[the remark this answers is not in the document]
Comment (nazarfil) [reply hexa/datasets/graphql/schema.graphql:430]: This can be , agree

File: hexa/datasets/queue.py (hunk 5 above)
Comment (nazarfil) [reply hexa/datasets/queue.py:53]: I though that setting default  to empty list for the model does already cover it

File: hexa/datasets/queue.py (hunk 3 above)
Comment (qgerome) [reply hexa/datasets/queue.py:50]: The problem with `head` (and we talked about that) is that it returns the first rows of the dataset and not a sample of it. What you get is then not representative of the data you'd have in it. 
I do not see in the documentation the limitation that it has to contain at least the number of records asked by the sample call to work.
If you look at the doc, you can see that you can set a seed to have a reproducible behaviour.

File: hexa/datasets/schema/queries.py (hunk 7 above)
Comment (qgerome) [reply hexa/datasets/schema/queries.py:49]: Let's start simple with a FK on `DatasetVersionFile` to `DatasetFileMetadata`

**Review by qgerome — CHANGES_REQUESTED** [review]

File: hexa/datasets/graphql/schema.graphql [hunk 10]
```diff
@@ -91,6 +91,24 @@ type DatasetVersionPermissions {
...
+    STATUS_FINISHED
+}
+
+"""
+Metadata for dataset file
+"""
+type DatasetFileMetadata {
+    content: JSON!
+    status: FileMetadataStatus!
+    statusReason: String
```
Comment (nazarfil) [inline hexa/datasets/graphql/schema.graphql:109]: to remove

File: hexa/datasets/graphql/schema.graphql [hunk 11]
```diff
@@ -429,6 +447,8 @@ extend type Query {
   dataset(id: ID!): Dataset
   "Get a dataset by its slug."
   datasetVersion(id: ID!): DatasetVersion
+  "Get a dataset file bi its id "
```
Comment (qgerome) [inline hexa/datasets/graphql/schema.graphql:450]: typo

File: hexa/datasets/models.py [hunk 12]
```diff
@@ -254,6 +256,33 @@ class Meta:
         ordering = ["uri"]
 
 
+class DatasetFileMetadata(Base):
+    STATUS_PROCESSING = "PROCESSING"
+    STATUS_FAILED = "FAILED"
+    STATUS_FINISHED = "FINISHED"
+
+    STATUS_CHOICES = [
+        (STATUS_PROCESSING, "Processing"),
```
Comment (qgerome) [inline hexa/datasets/models.py:265]: no gettext ?

File: hexa/datasets/queue.py [hunk 13]
```diff
@@ -0,0 +1,105 @@
...
+)
+
+logger = getLogger(__name__)
+
+
+def download_file_as_dataframe(
+    dataset_version_file: DatasetVersionFile,
+) -> pd.DataFrame:
+    filename = dataset_version_file.filename
+    file_format = filename.split(".")[-1]
```
Comment (qgerome) [inline hexa/datasets/queue.py:24]: It would make sense to use this https://docs.python.org/3/library/mimetypes.html

File: hexa/datasets/queue.py [hunk 14]
```diff
@@ -0,0 +1,105 @@
...
+    DatasetVersionFile,
+)
+
+logger = getLogger(__name__)
+
+
+def download_file_as_dataframe(
+    dataset_version_file: DatasetVersionFile,
+) -> pd.DataFrame:
+    filename = dataset_version_file.filename
```
Comment (qgerome) [inline hexa/datasets/queue.py:23]: not sure why you'd need to put that in another variable

File: hexa/datasets/schema/types.py [hunk 15]
```diff
@@ -217,6 +218,16 @@ def resolve_upload_url(obj, info, **kwargs):
         return None
 
 
+@dataset_version_file_object.field("fileMetadata")
+def resolve_version_file_metadata(obj: DatasetVersionFile, info, **kwargs):
+    try:
+        file_metadata = DatasetFileMetadata.objects.get(dataset_version_file=obj)
```
Comment (qgerome) [inline hexa/datasets/schema/types.py:224]: You can have plenty of DatasetFileMetadata per DatasetVersionFile (as specified by your FK)

File: hexa/datasets/schema/mutations.py [hunk 16]
```diff
@@ -252,11 +252,13 @@ def resolve_create_version_file(_, info, **kwargs):
                     uri=version.get_full_uri(mutation_input["uri"]),
                     content_type=mutation_input["contentType"],
                 )
-                return {
-                    "success": True,
-                    "errors": [],
-                    "file": file,
-                }
+
+            load_file_metadata(file_id=file.id)
```
Comment (qgerome) [inline hexa/datasets/schema/mutations.py:256]: I would add this as a method of the DatasetVersionFile. 
```
DatasetVersionFile.load_metadata()
```

File: hexa/datasets/schema/types.py (hunk 15 above)
Comment (qgerome) [reply hexa/datasets/schema/types.py:224]: I'd create a property on DatasetVersionFile to get the latest DatasetFileMetadata

**Review by qgerome — CHANGES_REQUESTED** [review]

File: hexa/datasets/models.py [hunk 17]
```diff
@@ -251,10 +253,41 @@ class DatasetVersionFile(Base):
...
+        choices=STATUS_CHOICES,
+        default=STATUS_PROCESSING,
+    )
+    status_reason = models.TextField(blank=True, null=True)
+    dataset_version_file = models.ForeignKey(
+        DatasetVersionFile,
+        null=False,
+        blank=False,
+        on_delete=models.CASCADE,
+        related_name="file_metadata",
```
Comment (qgerome) [inline hexa/datasets/models.py:287]: If the link will get you a list of items, better to put that in plural

**Review by qgerome — COMMENTED** [review]
Small typo and we are good to go !

**Comment by qgerome** [comment]
Also, I tried to upload a new dataset version file and it did not work.

```
app               | 
app               | Traceback (most recent call last):
app               |   File "/usr/local/lib/python3.11/site-packages/graphql/execution/execute.py", line 1030, in execute
app               |     result = exe_context.execute_operation(operation, root_value)
app               |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
app               |   File "/usr/local/lib/python3.11/site-packages/graphql/execution/execute.py", line 353, in execute_operation
app               |     return (
app               |            ^
app               |   File "/usr/local/lib/python3.11/site-packages/graphql/execution/execute.py", line 375, in execute_fields_serially
app               |     result = self.execute_field(
app               |              ^^^^^^^^^^^^^^^^^^^
app               |   File "/usr/local/lib/python3.11/site-packages/graphql/execution/execute.py", line 558, in execute_field
app               |     self.handle_field_error(error, return_type)
app               |   File "/usr/local/lib/python3.11/site-packages/graphql/execution/execute.py", line 569, in handle_field_error
app               |     raise error
app               |   File "/usr/local/lib/python3.11/site-packages/graphql/execution/execute.py", line 521, in execute_field
app               |     result = resolve_fn(source, info, **args)
app               |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
app               |   File "/code/hexa/user_management/schema.py", line 76, in resolve
app               |     return original_resolver(obj, info, **kwargs)
app               |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
app               |   File "/code/hexa/datasets/schema/mutations.py", line 256, in resolve_create_version_file
app               |     load_file_metadata(file_id=file.id)
app               |   File "/code/hexa/datasets/queue.py", line 107, in load_file_metadata
app               |     {
app               | graphql.error.graphql_error.GraphQLError: unhashable type: 'dict'
app               | 
app               | GraphQL request:2:3
app               | 1 | mutation CreateDatasetVersionFile($input: CreateDatasetVersionFileInput!) {
app               | 2 |   createDatasetVersionFile(input: $input) {
app               |   |   ^
app               | 3 |     uploadUrl
app               | [06/Aug/2024 13:24:55] "POST /graphql/DatasetVersionFilesDataGrid/ HTTP/1.1" 200 223
app               | [06/Aug/2024 13:24:55] "POST /graphql/CreateDatasetVersionFile/ HTTP/1.1" 200 937
app               | unhashable type: 'dict'
app               | 
app               | GraphQL request:2:3
app               | 1 | mutation CreateDatasetVersionFile($input: CreateDatasetVersionFileInput!) {
app               | 2 |   createDatasetVersionFile(input: $input) {
app               |   |   ^
app               | 3 |     uploadUrl
app               | Traceback (most recent call last):
app               |   File "/usr/local/lib/python3.11/site-packages/graphql/execution/execute.py", line 521, in execute_field
app               |     result = resolve_fn(source, info, **args)
app               |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
app               |   File "/code/hexa/user_management/schema.py", line 76, in resolve
app               |     return original_resolver(obj, info, **kwargs)
app               |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
app               |   File "/code/hexa/datasets/schema/mutations.py", line 256, in resolve_create_version_file
app               |     load_file_metadata(file_id=file.id)
app               |   File "/code/hexa/datasets/queue.py", line 107, in load_file_metadata
app               |     {
app               | TypeError: unhashable type: 'dict'
```

**Comment by nazarfil** [comment]
> Also, I tried to upload a new dataset version file and it did not work.
> 
> ```
> app               | 
> app               | Traceback (most recent call last):
> app               |   File "/usr/local/lib/python3.11/site-packages/graphql/execution/execute.py", line 1030, in execute
> app               |     result = exe_context.execute_operation(operation, root_value)
> app               |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> app               |   File "/usr/local/lib/python3.11/site-packages/graphql/execution/execute.py", line 353, in execute_operation
> app               |     return (
> app               |            ^
> app               |   File "/usr/local/lib/python3.11/site-packages/graphql/execution/execute.py", line 375, in execute_fields_serially
> app               |     result = self.execute_field(
> app               |              ^^^^^^^^^^^^^^^^^^^
> app               |   File "/usr/local/lib/python3.11/site-packages/graphql/execution/execute.py", line 558, in execute_field
> app               |     self.handle_field_error(error, return_type)
> app               |   File "/usr/local/lib/python3.11/site-packages/graphql/execution/execute.py", line 569, in handle_field_error
> app               |     raise error
> app               |   File "/usr/local/lib/python3.11/site-packages/graphql/execution/execute.py", line 521, in execute_field
> app               |     result = resolve_fn(source, info, **args)
> app               |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> app               |   File "/code/hexa/user_management/schema.py", line 76, in resolve
> app               |     return original_resolver(obj, info, **kwargs)
> app               |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> app               |   File "/code/hexa/datasets/schema/mutations.py", line 256, in resolve_create_version_file
> app               |     load_file_metadata(file_id=file.id)
> app               |   File "/code/hexa/datasets/queue.py", line 107, in load_file_metadata
> app               |     {
> app               | graphql.error.graphql_error.GraphQLError: unhashable type: 'dict'
> app               | 
> app               | GraphQL request:2:3
> app               | 1 | mutation CreateDatasetVersionFile($input: CreateDatasetVersionFileInput!) {
> app               | 2 |   createDatasetVersionFile(input: $input) {
> app               |   |   ^
> app               | 3 |     uploadUrl
> app               | [06/Aug/2024 13:24:55] "POST /graphql/DatasetVersionFilesDataGrid/ HTTP/1.1" 200 223
> app               | [06/Aug/2024 13:24:55] "POST /graphql/CreateDatasetVersionFile/ HTTP/1.1" 200 937
> app               | unhashable type: 'dict'
> app               | 
> app               | GraphQL request:2:3
> app               | 1 | mutation CreateDatasetVersionFile($input: CreateDatasetVersionFileInput!) {
> app               | 2 |   createDatasetVersionFile(input: $input) {
> app               |   |   ^
> app               | 3 |     uploadUrl
> app               | Traceback (most recent call last):
> app               |   File "/usr/local/lib/python3.11/site-packages/graphql/execution/execute.py", line 521, in execute_field
> app               |     result = resolve_fn(source, info, **args)
> app               |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> app               |   File "/code/hexa/user_management/schema.py", line 76, in resolve
> app               |     return original_resolver(obj, info, **kwargs)
> app               |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> app               |   File "/code/hexa/datasets/schema/mutations.py", line 256, in resolve_create_version_file
> app               |     load_file_metadata(file_id=file.id)
> app               |   File "/code/hexa/datasets/queue.py", line 107, in load_file_metadata
> app               |     {
> app               | TypeError: unhashable type: 'dict'
> ```

I remember fixing this one, will add fix in this branch

**Review by qgerome — APPROVED** [review]
